lib/Transforms/YosysOptimizer/YosysOptimizer.cpp

```diff
@@ -79,63 +82,234 @@ struct YosysOptimizer : public impl::YosysOptimizerBase<YosysOptimizer> {
-    std::error_code EC;
-    llvm::raw_fd_ostream of(filename, EC);
-    if (failed(translateToVerilog(op, of)) || EC) {
-      return WalkResult::interrupt();
+tensor::FromElementsOp convertIntegerValue(Value value, Type convertedType,
+                                           OpBuilder &b, Location loc) {
+  IntegerType argType = value.getType().cast<IntegerType>();
+  int width = argType.getWidth();
+  SmallVector<Value> extractedBits;
+  extractedBits.reserve(width);
+
+  for (int i = 0; i < width; i++) {
+    // These arith ops correspond to extracting the i-th bit
+    // from the input
+    auto shiftAmount =
+        b.create<arith::ConstantOp>(loc, argType, b.getIntegerAttr(argType, i));
+    auto bitMask = b.create<arith::ConstantOp>(
+        loc, argType, b.getIntegerAttr(argType, 1 << i));
+    auto andOp = b.create<arith::AndIOp>(loc, value, bitMask);
+    auto shifted = b.create<arith::ShRSIOp>(loc, andOp, shiftAmount);
+    extractedBits.push_back(
+        b.create<arith::TruncIOp>(loc, b.getI1Type(), shifted));
+  }
+
+  return b.create<tensor::FromElementsOp>(loc, convertedType,
+                                          ValueRange{extractedBits});
+}
+
+/// Convert a secret.generic's operands secret.secret<i3>
+/// to secret.secret<tensor<3xi1>>.
+LogicalResult convertOpOperands(secret::GenericOp op, func::FuncOp func,
+                                SmallVector<Value> &typeConvertedArgs) {
+  for (OpOperand &opOperand : op->getOpOperands()) {
+    Type convertedType =
+        func.getFunctionType().getInputs()[opOperand.getOperandNumber()];
+
+    if (!opOperand.get().getType().isa<secret::SecretType>()) {
+      // The type is not secret, but still must be booleanized
+      OpBuilder builder(op);
+      auto fromElementsOp = convertIntegerValue(opOperand.get(), convertedType,
+                                                builder, op.getLoc());
+      typeConvertedArgs.push_back(fromElementsOp.getResult());
+      continue;
+    }
+
+    secret::SecretType originalType =
+        opOperand.get().getType().cast<secret::SecretType>();
+    if (!originalType.getValueType().isa<IntegerType>()) {
+      op.emitError() << "Unsupported input type to secret.generic: "
+                     << originalType.getValueType();
+      return failure();
+    }
+
+    // Insert a conversion from the original type to the converted type
+    OpBuilder builder(op);
+    typeConvertedArgs.push_back(builder.create<secret::CastOp>(
+        op.getLoc(), secret::SecretType::get(convertedType), opOperand.get()));
+  }
+
+  return success();
+}
+
+/// Convert a secret.generic's results from secret.secret<tensor<3xi1>>
+/// to secret.secret<i3>.
+LogicalResult convertOpResults(secret::GenericOp op,
+                               DenseSet<Operation *> &castOps,
+                               SmallVector<Value> &typeConvertedResults) {
+  for (Value opResult : op.getResults()) {
+    if (!opResult.getType().isa<secret::SecretType>()) {
```

Comment:
could you add a testcase for this case of a non-secret yield? I'm having trouble understanding what happens (why doesn't the tensor.from_elements result need to be concatted back to an int?)

Reply:
Actually, I think this code is unreachable. `secret.generic` always has a secret result. This is enforced here: https://github.com/google/heir/blob/6e58f1d4686193262b1b52c572a64e57fddb972a/lib/Dialect/Secret/IR/SecretOps.cpp#L47-L53

I will remove this check and add a comment.